File: data-pipeline/imglab/core/metrics.py
"""Fidelity metrics for the offline pipeline, identical in formula to the browser twin
(frontend/src/engine/metrics.ts): PSNR on RGB MSE, SSIM (Wang et al. 2004, DOI 10.1109/TIP.2003.819861)
on luma with an 11x11 Gaussian window (sigma 1.5, reflect padding), and MS-SSIM (Wang et al. 2003) over
five scales. Inputs are float arrays in [0, 1], shape (H, W, 3) for RGB or (H, W) for a single channel.

Sanity (see tests): ssim(x, x) == 1, psnr(x, x) == inf, and a uniform 1/255 shift gives PSNR ~= 48.13 dB.
"""
from __future__ import annotations

import numpy as np
from scipy.ndimage import gaussian_filter

_LUMA = np.array([0.2126, 0.7152, 0.0722], dtype=np.float64)
_C1 = 0.01**2
_C2 = 0.03**2
_MSSSIM_WEIGHTS = np.array([0.0448, 0.2856, 0.3001, 0.2363, 0.1333], dtype=np.float64)
# truncate so the Gaussian radius is 5 (an 11-tap kernel), matching the browser implementation.
_TRUNCATE = 5.0 / 1.5


def to_luma(img: np.ndarray) -> np.ndarray:
    img = np.asarray(img, dtype=np.float64)
    if img.ndim == 2:
        return img
    return img[..., :3] @ _LUMA


def mse(a: np.ndarray, b: np.ndarray) -> float:
    a = np.asarray(a, dtype=np.float64)
    b = np.asarray(b, dtype=np.float64)
    return float(np.mean((a - b) ** 2))


def psnr(a: np.ndarray, b: np.ndarray) -> float:
    m = mse(a, b)
    if m <= 0:
        return float("inf")
    return float(10.0 * np.log10(1.0 / m))  # MAX = 1 for [0, 1] data


def _blur(x: np.ndarray) -> np.ndarray:
    return gaussian_filter(x, sigma=1.5, truncate=_TRUNCATE, mode="reflect")


def _ssim_terms(ya: np.ndarray, yb: np.ndarray):
    mu_a = _blur(ya)
    mu_b = _blur(yb)
    va = _blur(ya * ya) - mu_a * mu_a
    vb = _blur(yb * yb) - mu_b * mu_b
    cov = _blur(ya * yb) - mu_a * mu_b
    return mu_a, mu_b, va, vb, cov


def ssim(a: np.ndarray, b: np.ndarray) -> float:
    ya = to_luma(a)
    yb = to_luma(b)
    mu_a, mu_b, va, vb, cov = _ssim_terms(ya, yb)
    num = (2 * mu_a * mu_b + _C1) * (2 * cov + _C2)
    den = (mu_a**2 + mu_b**2 + _C1) * (va + vb + _C2)
    return float(np.mean(num / den))


def _cs(ya: np.ndarray, yb: np.ndarray) -> float:
    _, _, va, vb, cov = _ssim_terms(ya, yb)
    return float(np.mean((2 * cov + _C2) / (va + vb + _C2)))


def _downsample2(y: np.ndarray) -> np.ndarray:
    h, w = y.shape[0] - y.shape[0] % 2, y.shape[1] - y.shape[1] % 2
    y = y[:h, :w]
    return 0.25 * (y[0::2, 0::2] + y[1::2, 0::2] + y[0::2, 1::2] + y[1::2, 1::2])


def ms_ssim(a: np.ndarray, b: np.ndarray) -> float:
    ya = to_luma(a)
    yb = to_luma(b)
    if min(ya.shape) < 176:
        return ssim(ya, yb)
    prod = 1.0
    for s in range(5):
        if s < 4:
            cs = max(1e-8, _cs(ya, yb))
            prod *= cs ** _MSSSIM_WEIGHTS[s]
            ya = _downsample2(ya)
            yb = _downsample2(yb)
        else:
            last = max(1e-8, ssim(ya, yb))
            prod *= last ** _MSSSIM_WEIGHTS[s]
    return float(prod)


def bpp(num_bytes: int, width: int, height: int) -> float:
    """Bits per pixel, for rate-distortion curves."""
    return (num_bytes * 8.0) / float(width * height)

File: frontend/src/engine/metrics.ts
// Fidelity metrics, identical in formula to the offline Python (data-pipeline/imglab/core/metrics.py).
// Operate on ImagePlanes in [0,1]. PSNR on RGB MSE; SSIM (Wang et al. 2004, DOI 10.1109/TIP.2003.819861)
// on luma with an 11x11 Gaussian window; MS-SSIM (Wang et al. 2003) over 5 scales. Sanity: SSIM(x,x)=1,
// PSNR(x,x)=Infinity, and a 1-LSB (1/255) uniform shift gives PSNR ~= 48.13 dB.
import type { ImagePlanes } from './image';
import { luma } from './image';

export function mse(a: ImagePlanes, b: ImagePlanes): number {
  const n = a.w * a.h;
  let s = 0;
  for (let i = 0; i < n; i++) {
    s += (a.r[i] - b.r[i]) ** 2 + (a.g[i] - b.g[i]) ** 2 + (a.b[i] - b.b[i]) ** 2;
  }
  return s / (n * 3);
}

export function psnr(a: ImagePlanes, b: ImagePlanes): number {
  const m = mse(a, b);
  if (m <= 0) return Infinity;
  return 10 * Math.log10(1 / m); // MAX = 1 for [0,1] data
}

// --- separable Gaussian blur (reflect padding) on a single channel ---
function gaussKernel(sigma: number, radius: number): Float32Array {
  const k = new Float32Array(2 * radius + 1);
  let sum = 0;
  for (let i = -radius; i <= radius; i++) {
    const v = Math.exp(-(i * i) / (2 * sigma * sigma));
    k[i + radius] = v;
    sum += v;
  }
  for (let i = 0; i < k.length; i++) k[i] /= sum;
  return k;
}

const reflect = (i: number, n: number): number => {
  if (i < 0) return -i - 1 < n ? -i - 1 : 0;
  if (i >= n) return 2 * n - i - 1 >= 0 ? 2 * n - i - 1 : n - 1;
  return i;
};

function blur(src: Float32Array, w: number, h: number, k: Float32Array, radius: number): Float32Array {
  const tmp = new Float32Array(w * h);
  const out = new Float32Array(w * h);
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      let s = 0;
      for (let t = -radius; t <= radius; t++) s += k[t + radius] * src[y * w + reflect(x + t, w)];
      tmp[y * w + x] = s;
    }
  }
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      let s = 0;
      for (let t = -radius; t <= radius; t++) s += k[t + radius] * tmp[reflect(y + t, h) * w + x];
      out[y * w + x] = s;
    }
  }
  return out;
}

/** Mean SSIM on the luma channel with an 11x11 Gaussian (sigma 1.5). Returns a value in (-1, 1]. */
export function ssim(a: ImagePlanes, b: ImagePlanes): number {
  if (a.w !== b.w || a.h !== b.h) throw new Error('ssim: size mismatch');
  const w = a.w;
  const h = a.h;
  const ya = luma(a);
  const yb = luma(b);
  const radius = 5;
  const k = gaussKernel(1.5, radius);
  const C1 = 0.01 ** 2;
  const C2 = 0.03 ** 2;

  const muA = blur(ya, w, h, k, radius);
  const muB = blur(yb, w, h, k, radius);
  const n = w * h;
  const yaa = new Float32Array(n);
  const ybb = new Float32Array(n);
  const yab = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    yaa[i] = ya[i] * ya[i];
    ybb[i] = yb[i] * yb[i];
    yab[i] = ya[i] * yb[i];
  }
  const sAA = blur(yaa, w, h, k, radius);
  const sBB = blur(ybb, w, h, k, radius);
  const sAB = blur(yab, w, h, k, radius);

  let sum = 0;
  for (let i = 0; i < n; i++) {
    const ma = muA[i];
    const mb = muB[i];
    const va = sAA[i] - ma * ma;
    const vb = sBB[i] - mb * mb;
    const cov = sAB[i] - ma * mb;
    const num = (2 * ma * mb + C1) * (2 * cov + C2);
    const den = (ma * ma + mb * mb + C1) * (va + vb + C2);
    sum += num / den;
  }
  return sum / n;
}

// contrast-structure term only (for MS-SSIM intermediate scales)
function csMap(ya: Float32Array, yb: Float32Array, w: number, h: number): number {
  const radius = 5;
  const k = gaussKernel(1.5, radius);
  const C2 = 0.03 ** 2;
  const muA = blur(ya, w, h, k, radius);
  const muB = blur(yb, w, h, k, radius);
  const n = w * h;
  const yaa = new Float32Array(n);
  const ybb = new Float32Array(n);
  const yab = new Float32Array(n);
  for (let i = 0; i < n; i++) {
    yaa[i] = ya[i] * ya[i];
    ybb[i] = yb[i] * yb[i];
    yab[i] = ya[i] * yb[i];
  }
  const sAA = blur(yaa, w, h, k, radius);
  const sBB = blur(ybb, w, h, k, radius);
  const sAB = blur(yab, w, h, k, radius);
  let sum = 0;
  for (let i = 0; i < n; i++) {
    const va = sAA[i] - muA[i] * muA[i];
    const vb = sBB[i] - muB[i] * muB[i];
    const cov = sAB[i] - muA[i] * muB[i];
    sum += (2 * cov + C2) / (va + vb + C2);
  }
  return sum / n;
}

function downsample2(y: Float32Array, w: number, h: number): { y: Float32Array; w: number; h: number } {
  const w2 = w >> 1;
  const h2 = h >> 1;
  const out = new Float32Array(w2 * h2);
  for (let j = 0; j < h2; j++) {
    for (let i = 0; i < w2; i++) {
      const x = 2 * i;
      const yy = 2 * j;
      out[j * w2 + i] = 0.25 * (y[yy * w + x] + y[yy * w + x + 1] + y[(yy + 1) * w + x] + y[(yy + 1) * w + x + 1]);
    }
  }
  return { y: out, w: w2, h: h2 };
}

const MSSSIM_WEIGHTS = [0.0448, 0.2856, 0.3001, 0.2363, 0.1333];

/** MS-SSIM over 5 scales; needs at least ~176px on the short side, else falls back to single-scale SSIM. */
export function msSsim(a: ImagePlanes, b: ImagePlanes): number {
  if (Math.min(a.w, a.h) < 176) return ssim(a, b);
  let ya = luma(a);
  let yb = luma(b);
  let w = a.w;
  let h = a.h;
  let prod = 1;
  for (let s = 0; s < 5; s++) {
    if (s < 4) {
      const cs = Math.max(1e-8, csMap(ya, yb, w, h));
      prod *= Math.pow(cs, MSSSIM_WEIGHTS[s]);
      const da = downsample2(ya, w, h);
      const db = downsample2(yb, w, h);
      ya = da.y;
      yb = db.y;
      w = da.w;
      h = da.h;
    } else {
      // final scale: full luminance-contrast-structure SSIM
      const planesA: ImagePlanes = { r: ya, g: ya, b: ya, w, h };
      const planesB: ImagePlanes = { r: yb, g: yb, b: yb, w, h };
      const last = Math.max(1e-8, ssim(planesA, planesB));
      prod *= Math.pow(last, MSSSIM_WEIGHTS[s]);
    }
  }
  return prod;
}
